source/ota_mqtt.c

```diff
@@ -226,6 +226,12 @@ static OtaErr_t subscribeToJobNotificationTopics( const OtaAgentContext_t * pAge
     assert( pAgentCtx != NULL );
 
     /* Build and subscribe to the first topic. */
+
+    /* MISRA rule 21.6 prohibits use of all functions from stdio.h because they have undefined
+     * behavior. So snprintf, undefined behavior only occurs when the destination buffer and format
+     * buffer overlaps. We are using static buffers and format strings defined in this files, and
```

Comment:
Can you provide a reference for the assertion that undefined behavior is limited to overlaps and only overlaps?

Reply:
Quote from C standard 7.19.6.5,

The snprintf function is equivalent to fprintf, except that the output is written into
an array (specified by argument s) rather than to a stream. If n is zero, nothing is written,
and s may be a null pointer. Otherwise, output characters beyond the n-1st are
discarded rather than being written to the array, and a null character is written at the end
of the characters actually written into the array. **If copying takes place between objects
that overlap, the behavior is undefined**.

Reply:
Have a look at this.  It includes links to Annex J.2 of the standard where undefined behaviors are listed.
https://www.misra.org.uk/forum/viewtopic.php?t=1725

Reply:
What is the defined behavior for snprintf when the number of parameters do not match the tokens in the format specifier, also what if the types do not match?

Some references (this is a small subset of the full list, really why these functions are outlawed):
C99 section 7.19.6.1 - 
Par 2 "If there are insufficient arguments for the format, the behavior is undefined."
Par 4 "If a precision appears with any other conversion specifier, the behavior is undefined."
Par 6 "For g and G conversions, trailing zeros are not removed from the result. For other conversions, the behavior is undefined."
Par 6 "... if a precision is specified, the 0 flag is ignored. For other conversions, the behavior is undefined."
Par 7 "...If a length modifier appears with any conversion specifier other than as specified above, the behavior is undefined."
Par 9 "If a conversion specification is invalid, the behavior is undefined."
Par 9 "If any argument is
not the correct type for the corresponding conversion specification, the behavior is undefined."